File: Design1_Problem2.py
# Time Complexity : O(1) for getting Min and push and pop
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : None

# Your code here along with comments explaining your approach : a min value where I compare with the already existing minimum value and update it accordingly

#Optimized
import sys
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.min_val = sys.maxint

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        if val <= self.min_val:
            self.stack.append(self.min_val)
            self.min_val = val
        self.stack.append(val)

    def pop(self):
        """
        :rtype: None
        """
        if self.min_val == self.stack.pop():
            self.min_val = self.stack[-1]
            self.stack.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]
        
    def getMin(self):
        """
        :rtype: int
        """
        return self.min_val



#BruteForce
# class MinStack(object):

#     def __init__(self):
#         self.stack = []
#         self.min_stack = [sys.maxint]

#     def push(self, val):
#         """
#         :type val: int
#         :rtype: None
#         """
#         self.stack.append(val)
#         if self.min_stack[-1] > val:
#             self.min_stack.append(val)
#         else:
#             self.min_stack.append(self.min_stack[-1])

#     def pop(self):
#         """
#         :rtype: None
#         """
#         self.stack.pop()
#         # if top == self.min_stack[-1]:
#         self.min_stack.pop()

#     def top(self):
#         """
#         :rtype: int
#         """
#         return self.stack[-1]
        
#     def getMin(self):
#         """
#         :rtype: int
#         """
#         return self.min_stack[-1] if self.min_stack[-1] != sys.maxint else 0


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Design1_Problem2.java
import java.util.Stack;

// Time Complexity : Push, Pop, getMin() : O(1)
// Space Complexity : O(1)


// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
/*
When min from Stack will be at top and if we pop an element then it caused the wrong reuslt for getMin().
Then I pushed the min element if the new element to insert is less than current min element. So, whenever Min is at top and if we pop that element we don't get the wrong output

*/


// Your code here along with comments explaining your approach

/*
Created stack using java.util inbuilt stack and one min variable to get min from stack in O(1) in future.

For push() : As I said in my problem which is fixed now, I pushed the min element if the new element to insert is less than current min element. 
If the new value is not less than current min then simply pushed into stack

For pop() : Simply popped the element from stack. If this element is min then again pop element and set that as min element.

For top() : Element stored at top returned using peek() of Stack.

For getMin() : Returning the value of min variable.
*/



class SplStack {
    Stack<Integer> stack;
    int min;
    public SplStack() {
        stack = new Stack<Integer>();
        min = Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        
        if(val <= min){
            stack.push(min);
            min = val;
        }
        stack.push(val);
    }
    
    public void pop() {
        int element = stack.pop();
        if(element == min){
            min = stack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min;
    }
}


public class Design1_Problem2{
    public static void main(String args[]){
        SplStack stack = new SplStack();
        stack.push(-2); stack.push(0); stack.push(-3);

        System.out.println("GetMin() : " + stack.getMin());

        stack.pop();
        System.out.println("Top() : " + stack.top());
        System.out.println("GetMin() : " + stack.getMin());

    }
}
